runtime/protocol/http-client/jvm/src/aws/smithy/kotlin/runtime/http/interceptors/requestcompression/GzipJvm.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.http.interceptors.requestcompression
+
+import aws.smithy.kotlin.runtime.ClientException
+import aws.smithy.kotlin.runtime.http.HttpBody
+import aws.smithy.kotlin.runtime.http.toHttpBody
+import aws.smithy.kotlin.runtime.io.SdkBuffer
+import aws.smithy.kotlin.runtime.io.SdkByteReadChannel
+import aws.smithy.kotlin.runtime.io.SdkSource
+import java.io.ByteArrayOutputStream
+import java.util.zip.GZIPOutputStream
+
+public actual class Gzip actual constructor() : CompressionAlgorithm {
+    actual override val id: String
+        get() = "gzip"
+
+    actual override suspend fun compress(stream: HttpBody): HttpBody = when (stream) {
+        is HttpBody.SourceContent -> GzipSdkSource(stream.readFrom()).toHttpBody()
+        is HttpBody.ChannelContent -> GzipByteReadChannel(stream.readFrom()).toHttpBody()
+        is HttpBody.Bytes -> compressByteArray(stream.bytes())
+        is HttpBody.Empty -> stream
+        else -> throw ClientException("HttpBody type is not supported")
+    }
+}
+
+internal fun compressByteArray(bytes: ByteArray): HttpBody {
+    val byteArrayOutputStream = ByteArrayOutputStream()
+    val gzipOutputStream = GZIPOutputStream(byteArrayOutputStream)
+
+    gzipOutputStream.write(bytes)
+    gzipOutputStream.close()
+
+    val compressedBody = byteArrayOutputStream.toByteArray().toHttpBody()
+    byteArrayOutputStream.close()
+
+    return compressedBody
+}
+
+internal class GzipSdkSource(
+    private val source: SdkSource,
+) : SdkSource by source {
+    private val gzipBuffer = SdkBuffer()
+    private val gzipOutputStream = GZIPOutputStream(gzipBuffer.outputStream())
+
+    override fun read(sink: SdkBuffer, limit: Long): Long { // TODO: Fix this
```

Comment:
**suggestion**: Rather than comment everywhere I propose the following (untested but should give you a general idea of the overall structure/flow):


```kotlin
private class GzipSdkSource (
    private val source: SdkSource
): SdkSource {
    private val gzipBuffer = SdkBuffer()
    private val gzipOutputStream = GZIPOutputStream(gzipBuffer.outputStream())

    override fun read(sink: SdkBuffer, limit: Long): Long {
        require(limit >= 0L)
        if (limit == 0L) return 0L

        val temp = SdkBuffer()
        val rc = source.read(temp, limit)

        if (rc == -1L) {
            // may trigger additional bytes written by gzip defalter
            gzipOutputStream.close()
        }

        // source is exhausted and nothing left buffered we are done
        if (rc == -1L && gzipBuffer.exhausted()) return -1L

        // compress what we read and add it to the buffer
        if (rc >= 0L) {
            gzipOutputStream.write(temp.readByteArray())
        }

        // read bytes read from compressed content
        return gzipBuffer.read(sink, limit)
    }

    override fun close() {
        source.close()
    }
}
```

The `SdkByteReadChannel` will look similar.